File: examples/public-api/javascript/02-streaming-execution.ts
/**
 * Example 02: Streaming Workflow Execution
 *
 * This example demonstrates real-time execution tracking using Server-Sent Events (SSE):
 * 1. Execute a workflow
 * 2. Stream execution events in real-time
 * 3. Display progress as nodes complete
 *
 * Run: npx tsx 02-streaming-execution.ts
 */

import "dotenv/config";
import { FlowMaestroClient } from "@flowmaestro/sdk";
import type { ExecutionEvent } from "@flowmaestro/sdk";

// Configuration from environment
const API_KEY = process.env.FLOWMAESTRO_API_KEY!;
const BASE_URL = process.env.FLOWMAESTRO_BASE_URL;
const WORKFLOW_ID = process.env.WORKFLOW_ID!;

// Progress bar helper
function progressBar(current: number, total: number, width: number = 30): string {
    const filled = Math.round((current / total) * width);
    const empty = width - filled;
    return `[${"=".repeat(filled)}${" ".repeat(empty)}] ${current}/${total}`;
}

async function main() {
    // Validate configuration
    if (!API_KEY || API_KEY === "fm_live_your_api_key_here") {
        console.error("Error: Please set FLOWMAESTRO_API_KEY in your .env file");
        process.exit(1);
    }
    if (!WORKFLOW_ID || WORKFLOW_ID === "wf_your_workflow_id") {
        console.error("Error: Please set WORKFLOW_ID in your .env file");
        process.exit(1);
    }

    const client = new FlowMaestroClient({
        apiKey: API_KEY,
        baseUrl: BASE_URL
    });

    console.log("FlowMaestro Streaming Execution Example\n");
    console.log("=".repeat(50));

    try {
        // Step 1: Execute the workflow
        console.log("\n1. Starting workflow execution...");
        const { data: execution } = await client.workflows.execute(WORKFLOW_ID, {
            inputs: {
                name: "Jane Smith",
                email: "jane@example.com"
            }
        });
        console.log(`   Execution ID: ${execution.execution_id}`);

        // Step 2: Stream execution events
        console.log("\n2. Streaming events (real-time):\n");

        let nodeCount = 0;
        let completedNodes = 0;
        const startTime = Date.now();

        // Using async iterator for cleaner code
        for await (const event of client.executions.streamIterator(execution.execution_id)) {
            displayEvent(event);

            // Track node progress
            if (event.type === "node:started") {
                nodeCount++;
            } else if (event.type === "node:completed") {
                completedNodes++;
                if (nodeCount > 0) {
                    console.log(`   Progress: ${progressBar(completedNodes, nodeCount)}`);
                }
            }

            // Check for terminal events
            if (isTerminalEvent(event)) {
                break;
            }
        }

        const duration = ((Date.now() - startTime) / 1000).toFixed(2);
        console.log(`\n   Total time: ${duration}s`);

        // Step 3: Fetch final execution state
        console.log("\n3. Final execution state:");
        const { data: finalState } = await client.executions.get(execution.execution_id);
        console.log(`   Status: ${finalState.status}`);

        if (finalState.outputs) {
            console.log("   Outputs:", JSON.stringify(finalState.outputs, null, 2));
        }

        console.log("\n" + "=".repeat(50));
        console.log("Streaming example completed!");
    } catch (error) {
        console.error("\nError:", error instanceof Error ? error.message : error);
        process.exit(1);
    }
}

function displayEvent(event: ExecutionEvent): void {
    const timestamp = new Date().toISOString().split("T")[1].slice(0, 12);
    const icon = getEventIcon(event.type);

    switch (event.type) {
        case "connected":
            console.log(`   ${timestamp} ${icon} Connected to execution stream`);
            break;

        case "execution:started":
            console.log(`   ${timestamp} ${icon} Execution started`);
            break;

        case "execution:progress":
            if (event.progress !== undefined) {
                console.log(`   ${timestamp} ${icon} Progress: ${event.progress}%`);
            }
            break;

        case "node:started":
            console.log(
                `   ${timestamp} ${icon} Node started: ${event.node_id} (${event.node_type || "unknown"})`
            );
            break;

        case "node:completed":
            console.log(`   ${timestamp} ${icon} Node completed: ${event.node_id}`);
            break;

        case "node:failed":
            console.log(`   ${timestamp} ${icon} Node failed: ${event.node_id} - ${event.error}`);
            break;

        case "execution:completed":
            console.log(`   ${timestamp} ${icon} Execution completed successfully`);
            break;

        case "execution:failed":
            console.log(`   ${timestamp} ${icon} Execution failed: ${event.error}`);
            break;

        case "execution:cancelled":
            console.log(`   ${timestamp} ${icon} Execution cancelled`);
            break;

        default:
            console.log(`   ${timestamp} ${icon} ${event.type}`);
    }
}

function getEventIcon(type: string): string {
    const icons: Record<string, string> = {
        connected: "[*]",
        "execution:started": "[>]",
        "execution:progress": "[~]",
        "node:started": "[+]",
        "node:completed": "[v]",
        "node:failed": "[x]",
        "execution:completed": "[V]",
        "execution:failed": "[X]",
        "execution:cancelled": "[-]"
    };
    return icons[type] || "[?]";
}

function isTerminalEvent(event: ExecutionEvent): boolean {
    return (
        event.type === "execution:completed" ||
        event.type === "execution:failed" ||
        event.type === "execution:cancelled"
    );
}

main();

File: examples/public-api/python/02_streaming_execution.py
"""
Example 02: Streaming Workflow Execution

This example demonstrates real-time execution tracking using Server-Sent Events (SSE):
1. Execute a workflow
2. Stream execution events in real-time
3. Display progress as nodes complete

Run: python 02_streaming_execution.py
"""

import os
import sys
import time
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configuration
API_KEY = os.getenv("FLOWMAESTRO_API_KEY")
BASE_URL = os.getenv("FLOWMAESTRO_BASE_URL")
WORKFLOW_ID = os.getenv("WORKFLOW_ID")

# Event icons for display
EVENT_ICONS = {
    "connected": "[*]",
    "execution:started": "[>]",
    "execution:progress": "[~]",
    "node:started": "[+]",
    "node:completed": "[v]",
    "node:failed": "[x]",
    "execution:completed": "[V]",
    "execution:failed": "[X]",
    "execution:cancelled": "[-]",
}

TERMINAL_EVENTS = {
    "execution:completed",
    "execution:failed",
    "execution:cancelled",
}


def progress_bar(current: int, total: int, width: int = 30) -> str:
    """Create a simple progress bar."""
    if total == 0:
        return f"[{' ' * width}] 0/0"
    filled = int((current / total) * width)
    empty = width - filled
    return f"[{'=' * filled}{' ' * empty}] {current}/{total}"


def format_timestamp() -> str:
    """Get current timestamp for display."""
    from datetime import datetime
    return datetime.now().strftime("%H:%M:%S.%f")[:-3]


def display_event(event: dict) -> None:
    """Display an event in a formatted way."""
    event_type = event.get("type", "unknown")
    icon = EVENT_ICONS.get(event_type, "[?]")
    timestamp = format_timestamp()

    if event_type == "connected":
        print(f"   {timestamp} {icon} Connected to execution stream")
    elif event_type == "execution:started":
        print(f"   {timestamp} {icon} Execution started")
    elif event_type == "execution:progress":
        progress = event.get("progress", 0)
        print(f"   {timestamp} {icon} Progress: {progress}%")
    elif event_type == "node:started":
        node_id = event.get("node_id", "unknown")
        node_type = event.get("node_type", "unknown")
        print(f"   {timestamp} {icon} Node started: {node_id} ({node_type})")
    elif event_type == "node:completed":
        node_id = event.get("node_id", "unknown")
        print(f"   {timestamp} {icon} Node completed: {node_id}")
    elif event_type == "node:failed":
        node_id = event.get("node_id", "unknown")
        error = event.get("error", "unknown")
        print(f"   {timestamp} {icon} Node failed: {node_id} - {error}")
    elif event_type == "execution:completed":
        print(f"   {timestamp} {icon} Execution completed successfully")
    elif event_type == "execution:failed":
        error = event.get("error", "unknown")
        print(f"   {timestamp} {icon} Execution failed: {error}")
    elif event_type == "execution:cancelled":
        print(f"   {timestamp} {icon} Execution cancelled")
    else:
        print(f"   {timestamp} {icon} {event_type}")


def main():
    # Validate configuration
    if not API_KEY or API_KEY == "fm_live_your_api_key_here":
        print("Error: Please set FLOWMAESTRO_API_KEY in your .env file")
        sys.exit(1)
    if not WORKFLOW_ID or WORKFLOW_ID == "wf_your_workflow_id":
        print("Error: Please set WORKFLOW_ID in your .env file")
        sys.exit(1)

    from flowmaestro import FlowMaestroClient

    client = FlowMaestroClient(
        api_key=API_KEY,
        base_url=BASE_URL if BASE_URL else None
    )

    print("FlowMaestro Streaming Execution Example\n")
    print("=" * 50)

    try:
        # Step 1: Execute the workflow
        print("\n1. Starting workflow execution...")
        response = client.workflows.execute(
            WORKFLOW_ID,
            inputs={
                "name": "Jane Smith",
                "email": "jane@example.com"
            }
        )
        execution = response["data"]
        execution_id = execution["execution_id"]
        print(f"   Execution ID: {execution_id}")

        # Step 2: Stream execution events
        print("\n2. Streaming events (real-time):\n")

        node_count = 0
        completed_nodes = 0
        start_time = time.time()

        for event in client.executions.stream(execution_id):
            display_event(event)

            # Track node progress
            event_type = event.get("type", "")
            if event_type == "node:started":
                node_count += 1
            elif event_type == "node:completed":
                completed_nodes += 1
                if node_count > 0:
                    print(f"   Progress: {progress_bar(completed_nodes, node_count)}")

            # Check for terminal events
            if event_type in TERMINAL_EVENTS:
                break

        duration = time.time() - start_time
        print(f"\n   Total time: {duration:.2f}s")

        # Step 3: Fetch final execution state
        print("\n3. Final execution state:")
        response = client.executions.get(execution_id)
        final_state = response["data"]
        print(f"   Status: {final_state['status']}")

        if final_state.get("outputs"):
            import json
            print(f"   Outputs: {json.dumps(final_state['outputs'], indent=2)}")

        print("\n" + "=" * 50)
        print("Streaming example completed!")

    except Exception as e:
        print(f"\nError: {e}")
        sys.exit(1)
    finally:
        client.close()


if __name__ == "__main__":
    main()
